src/overlays/effects/ovl_Effect_Ss_Solder_Srch_Ball/z_eff_ss_solder_srch_ball.c

```diff
@@ -1,6 +1,55 @@
-#include <ultra64.h>
-#include <global.h>
+/*
+ * File: z_eff_ss_solder_srch_ball.c
+ * Overlay: ovl_Effect_Ss_Solder_Srch_Ball
+ * Description: Vision sphere for courtyard guards
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/effects/ovl_Effect_Ss_Solder_Srch_Ball/func_809AC130.s")
+#include "z_eff_ss_solder_srch_ball.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/effects/ovl_Effect_Ss_Solder_Srch_Ball/func_809AC1AC.s")
+u32 EffSsSolderSrchBall_Init(GlobalContext* globalCtx, u32 index, EffectSs* this,
+                             EffSsSolderSrchBallInitParams* initParams);
+void EffSsSolderSrchBall_Update(GlobalContext* globalCtx, u32 index, EffectSs* this);
+
+EffectSsInit Effect_Ss_Solder_Srch_Ball_InitVars = {
+    EFFECT_SS_SOLDER_SRCH_BALL,
+    EffSsSolderSrchBall_Init,
+};
+
+u32 EffSsSolderSrchBall_Init(GlobalContext* globalCtx, u32 index, EffectSs* this,
+                             EffSsSolderSrchBallInitParams* initParams) {
+    this->pos = initParams->pos;
+    this->velocity = initParams->velocity;
+    this->accel = initParams->accel;
+    this->update = EffSsSolderSrchBall_Update;
+    this->life = 100;
+    this->regs[1] = initParams->unk_24;
+    this->unk_3C = initParams->linkDetected;
+    return 1;
+}
+
+void EffSsSolderSrchBall_Update(GlobalContext* globalCtx, u32 index, EffectSs* this) {
+    s32 pad;
+    f32 playerPosDiffX;
+    f32 playerPosDiffY;
+    f32 playerPosDiffZ;
+    s16* linkDetected;
+    Player* player = PLAYER;
+
+    linkDetected = this->unk_3C;
+
+    playerPosDiffX = player->actor.posRot.pos.x - this->pos.x;
+    playerPosDiffY = player->actor.posRot.pos.y - this->pos.y;
+    playerPosDiffZ = player->actor.posRot.pos.z - this->pos.z;
+
+    // if the sphere does not hit a wall (something blocks it) and is within 70 units of link, he has been detected
+    if (!func_8003E30C(&globalCtx->colCtx, &this->pos, 30.0f)) {
+        if (sqrtf(SQ(playerPosDiffX) + SQ(playerPosDiffY) + SQ(playerPosDiffZ)) < 70.0f) {
+            *linkDetected = true;
+        }
+    } else {
+        // dont let life reach -1 so that it keeps going until it hits a wall or gets blocked
+        if (this->life > 1) {
+            this->life = 1;
+        }
```

Comment:
I don't understand this comment. How is this code supposed to prevent life from reaching `-1`?

Reply:
yeah turns out i dont understand what the bgcheck func is doing at all. thought it was returning true if it hit a wall. oh well, took both comments out for now